### DISCUSSION TITLE
Have there been any attempts to revive this?

### DISCUSSION THREAD

**Discussion by tsbertalan** [tud-cor/FS19_modROS#54] [Q&A]
When evaluating both FS19 and FS22 as simulation platforms, I found that the popular AutoDrive ([FS22 version](https://github.com/Stephan-S/FS22_AutoDrive) or [19](https://github.com/Stephan-S/FS19_AutoDrive)) and Courseplay ([22](https://github.com/Courseplay/Courseplay_FS22) or [19](https://github.com/Courseplay/Courseplay)) mods have switched all recent development to FS22. In addition, ROS1 will be end-of-life this y ear.

For my purposes, a full ROS integration is not necessary, but it would probably be the easiest way, and would avoid duplicating effort. Do the authors of this package know of any ongoing efforts in this vein? If not, I might pull a few components such as the lua socket bridge for my own use, but ROS integration would be a side effort.

**Comment by gavanderhoorn** [comment]
There has been no new work on our side.

FS has fundamental limitations which make a proper ROS integration/plugin basically impossible/very difficult.

We've not looked at FS22, but I personally don't expect things to have changed significantly. Do you have any insight into how plugins/mods are created in FS22? If it's still Lua in a sandboxed VM with no networking capabilities, nor ability to run native code, we'd be working against the same restrictions as in FS19.

> For my purposes, a full ROS integration is not necessary, but it would probably be the easiest way, and would avoid duplicating effort.

we would also still like to see a better/more complete ROS mode for FS (no matter the version).

> If not, I might pull a few components such as the lua socket bridge for my own use, but ROS integration would be a side effort.

I'd be willing to support anyone trying to improve / continue the mod, but it would have to be with the goal of creating a ROS mod. All the code is `Apache-2.0` of course, but it would be nice if we could keep development centralised, instead of fragmenting the already (*very*) small community of ROS-with-FS.

PS:

> ROS1 will be end-of-life this year.

According to [REP-3: Target Platforms](https://www.ros.org/reps/rep-0003.html#noetic-ninjemys-may-2020-may-2025), Noetic support will end May, 2025. That's not end of this year.

**Reply by gavanderhoorn** [reply]
Friendly ping @tsbertalan

**Reply by tsbertalan** [reply]
Sorry, your first message got lost in the ether. Yes, it's Lua (and the "documentation" seems to be just a doxygen dump of function parameter names, with little commentary), though I haven't yet tried any networking. My plan was to have a companion Python or c++ program that would do the real work, and then use the lightest method feasible to connect it to the Lua. Probably IPC sockets (as I thought you did?) not full-blown networking, though I don't have much experience with either. The only output I yet achieved is writing some coordinates to a csv file. I don't know if that was possible in 19 or not--maybe it indicates a weakening of their sandbox?

Honestly, I've only done a little prototyping on my own repo for this project, so I'd be happy to take what I learn from that and make an FS22 fork of your repo, if you think that's the best way to go about it.

I'm traveling right now with limited access to my FS machine to try things out, but am still eager to hear what you think about this.

**Comment by tsbertalan** [comment]
I'm doing some more research, and thinking that doing some work of my own to port this to FS22 might be worthwhile, although it would be as a slow hobby project.

>  If it's still Lua in a sandboxed VM with no networking capabilities, nor ability to run native code, we'd be working against the same restrictions as in FS19.

I started trying to send some messages from an FS22 mod to Python by socket, and then realized I couldn't do `local socket = require("socket")`. (FWIW, I also looked at [protobuf](https://github.com/starwing/lua-protobuf) a bit, but I don't expect that would work any better, though it might give us a speedup.)

Then (should have done this first), I started reading through the code in this repo some more, and I have a few questions, that I'll separate into categories.

First, about Lua <--> Python IPC (the Python <--> ROS part is clear enough).

- [x] Is the lua --> python in fact accomplished by writing to a file on disk?
  - I'm not sure what the `io.open` [here](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/ros/WriteOnlyFileConnection.lua#LL49C16-L49C23) does, but that seems reasonable to assume.
  - I guess this is then symlinked to the same folder as the Windows python script, which then [opens it](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/all_in_one_publisher.py#L99) with `win32pipe.CreateNamedPipe`.
  - [This discussion](https://forum.giants-software.com/viewtopic.php?t=182643) on someone's "telemetry" mod seems related, though I believe [they used the same approach as you](https://github.com/Marciel032/FarmingSimulatorTelemetry/blob/7b93c844ab52695b15135f35ea558ee2e2ff8897/TelemetriaFarmingSimulator/FSTelemetry.lua#L588).
- [x]  Then, is the python --> Lua done by memory modification?
  - It seems that [`memorpy.MemWorker`](https://github.com/rbcastner/memorpy3) (not sure if that's the right repo) finds beginning and end sentinels on either end of a 64-byte region of memory in [`FarmingSimulator2019Game.exe`](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/cm_vel_subscriber.py#L38) and provides methods to modify that memory.
- [x] Why this particular division of responsibilities?
  - There are [read](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/cm_vel_subscriber.py#L108), write, and clear methods from the Python, side, but it seems that only [reading](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/modROS.lua#L223) is used from the Lua side.
  - Why do we need both the file pipe and the shared memory? Is it just because laserscans are big and twists are small (specifically, less than 64 bytes)?

Broader modding concerns:

- [x] Are there any high-level tutorials better than the bare doxygen dump Giants provides?
  - Reading through this repo, it does seem largely similar to my basic understanding of FS22 modding (modDesc.xml; specializations, ...), but just learning their API by reading various FS22 (or FS19) mods is not ideal.
  - It's a terrible rat's nest of global functions, implied callbacks that modder needs to define (but that won't necessarily raise errors if missing), and XML-as-code (e.g. `extraSourceFiles` is really an `exec` in disguise).
  - This is mostly just a rant.
- [x] What did you find (at least in FS19) were the available methods for exerting control on a vehicle or vehicles?
  - In FS22 there seem to be methods to either
    - specify directly the curvature of turn to be followed (for on-rails movement like the AI road traffic),
    - or to provide a number in time units, which I guess is something of a turn-angle target, which the vehicle will then start moving toward at constant vehicle-dependent rate.
    - It's pretty mysterious to me why this is in time units.
  - But I didn't see any way to provide a steering torque command, which would be desirable from the point of view of creating a realistic simulation interface.
  - For the engine commands, I remember it being less fraught, but I forget how--I haven't yet tried these methods.
- [x] Did you see any way to get multiple camera views at once?
  - I see that you used d3dshot and a fixed camera angle to provide camera output--I've been doing some of this myself.
  - However, [this forum post](https://gdn.giants-software.com/thread.php?categoryId=3&threadId=709) suggests that it might be possible to render multiple camera views into screen space at once, though that was 2011 I can't find any evidence that anyone actually did this. Does it seem possible to you?

**Reply by gavanderhoorn** [reply]
>> If it's still Lua in a sandboxed VM with no networking capabilities, nor ability to run native code, we'd be working against the same restrictions as in FS19.
> 
> I started trying to send some messages from an FS22 mod to Python by socket, and then realized I couldn't do `local socket = require("socket")`.

:)

yes, exactly. That's what I meant when I wrote "sandboxed VM with no networking capabilities".

> (FWIW, I also looked at [protobuf](https://github.com/starwing/lua-protobuf) a bit, but I don't expect that would work any better, though it might give us a speedup.)

without a way to get data out of the FS Lua sandbox, it wouldn't matter which (de)serialisation format you'd use I believe, as the bytes can't really go anywhere.

> Is the lua --> python in fact accomplished by writing to a file on disk?

It's an NTFS symlink which redirects a file handle to a named pipe.

Initial testing (but that may not have been thorough enough, seeing the telemetry mod you refer below) showed that FS19 only supported directly opening files in very specific places. The Lua interpreter has all sorts of filters in place to prevent mod authors from escaping the sandbox, so we used the symlink redirect to in essence create a sort of "infinite file". The Python scripts then just interact with the named pipe.

> I'm not sure what the `io.open` [here](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/ros/WriteOnlyFileConnection.lua#LL49C16-L49C23) does, but that seems reasonable to assume.

it basically just opens a file in the mod's base dir, which is the same file that's symlinked to the named pipe.

> I guess this is then symlinked to the same folder as the Windows python script, which then [opens it](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/all_in_one_publisher.py#L99) with `win32pipe.CreateNamedPipe`.

~~folder~~ -> *file*, but yes, that's the idea.

> [This discussion](https://forum.giants-software.com/viewtopic.php?t=182643) on someone's "telemetry" mod seems related, though I believe [they used the same approach as you](https://github.com/Marciel032/FarmingSimulatorTelemetry/blob/7b93c844ab52695b15135f35ea558ee2e2ff8897/TelemetriaFarmingSimulator/FSTelemetry.lua#L588).

Seems so. It seems slightly less cumbersome though, as they seem to directly open the named pipe from FS. Our initial testing seemed to imply that wasn't allowed / supported, which is why we went with the symlink approach.

Would be a nice 'final' update to remove that, as it would make setting up and using `modROS` a lot easier.

> Then, is the python --> Lua done by memory modification?

yes. As also [mentioned in the readme](https://github.com/tud-cor/FS19_modROS/tree/869a170967091ee8a9ec7b833ad7aaa06d3f3828#%EF%B8%8F-warning-and-disclaimer).

The named pipe -- at least the way we implemented it -- is uni-directional.

> It seems that [`memorpy.MemWorker`](https://github.com/rbcastner/memorpy3) (not sure if that's the right repo)

Unfortunately `memorpy` is/was in a bit of flux, so we had to apply some patches. We use(d) this fork: [tud-cor/memorpy3@wip_py3_patches](https://github.com/tud-cor/memorpy3/tree/wip_py3_patches).

> finds beginning and end sentinels on either end of a 64-byte region of memory in [`FarmingSimulator2019Game.exe`](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/cm_vel_subscriber.py#L38) and provides methods to modify that memory.
> 
> Why this particular division of responsibilities?

I'm not sure I follow completely.

Are you asking why a pipe *and* memory modification? If so: see my earlier comment.

> There are [read](https://github.com/tud-cor/fs_mod_ros_windows/blob/467ad87d855b5e4a7acc2de86e4a685e278b9a24/nodes/cm_vel_subscriber.py#L108), write, and clear methods from the Python, side, but it seems that only [reading](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/modROS.lua#L223) is used from the Lua side.

Lua strings are immutable. "Writing" to them will create a new instance/string, which we would have to track on the Python side (not impossible, but too cumbersome).

The idea was to create a single, large, string on the Lua side and use that as a piece of shared memory. Then overlay circular buffer semantics (or something like that), and use it as a socket/pipe.

As you'll see me commenting below quite a few times: we never got to that point.

(the memory modification functionality is far from finished. I have a fork/branch somewhere with code which is a bit more polished)

> Why do we need both the file pipe and the shared memory? Is it just because laserscans are big and twists are small (specifically, less than 64 bytes)?

See again my earlier comments: the pipe is/was unidirectional, so we needed a different entrypoint into FS to stream in data.

As the socket route was blocked, memory modification of (large) string buffers seemed like an approach with potential.
We never really got to the point where all of that was integrated completely and had feature-parity with regular sockets though.

> Are there any high-level tutorials better than the bare doxygen dump Giants provides?

there are, but IIRC, not from Giant themselves.

As you already mention below, modding Giant Lua-based games is a bit *messy*. There is a huge community -- and some of the community members eventually even became Giant employees, partly (?) due to their modding skills -- but it's fragmented, and hard to really get into.

I seem to remember there are some books available, but that's either *very* high-level, and any interesting mod quickly starts to need information at much deeper level, or it focuses on things like UIs, which, while important, were not what `modROS` needed.

> Reading through this repo, it does seem largely similar to my basic understanding of FS22 modding (modDesc.xml; specializations, ...), but just learning their API by reading various FS22 (or FS19) mods is not ideal.

That's exactly how we felt when we started working on this.

> It's a terrible rat's nest of global functions, implied callbacks that modder needs to define (but that won't necessarily raise errors if missing), and XML-as-code (e.g. `extraSourceFiles` is really an `exec` in disguise).
> This is mostly just a rant.

yes, but a justified one I believe.

Unfortunately not something I can really do anything about.

`modROS` doesn't do that much better when it comes to documentation, though we tried.


> What did you find (at least in FS19) were the available methods for exerting control on a vehicle or vehicles?
>       
> * In FS22 there seem to be methods to either
>    * specify directly the curvature of turn to be followed (for on-rails movement like the AI road traffic),
>    * or to provide a number in time units, which I guess is something of a turn-angle target, which the vehicle will then start moving toward at constant vehicle-dependent rate.
>    * It's pretty mysterious to me why this is in time units.

that's about what we found as well I believe -- it's been some time I last looked into this.

Some of the 'automation' mods for FS19 provided some inspiration.

> * But I didn't see any way to provide a steering torque command, which would be desirable from the point of view of creating a realistic simulation interface.

yes, so here we get to an important point I believe.

`modROS` was a PoC to see whether FS could be used as a "realistic simulator".

While we showed integration of external interfaces is somewhat possible, there are fundamental limitations in how FS works and Lua is integrated with it which seem to make it unsuitable for the kinds of simulations we had in mind.

Giants does / did have a *serious games* department IIRC, and FS has been used as training material for heavy machinery operators/drivers, but that kind of integration work can only really be done by Giants. You need access to engine and game internals, and the Lua API -- at least the parts that are (partially) documented -- doesn't offer that.

> * Did you see any way to get multiple camera views at once?
>   * I see that you used d3dshot and a fixed camera angle to provide camera output--I've been doing some of this myself.
>   * However, [this forum post](https://gdn.giants-software.com/thread.php?categoryId=3&threadId=709) suggests that it might be possible to render multiple camera views into screen space at once, though that was 2011 I can't find any evidence that anyone actually did this. Does it seem possible to you?

You can render multiple views, but IIRC, only for use as 'mirrors' (or virtual mirrors, ie: camera based mirrors).

So in order to be able to actually look at those multiple views, they'd have to be on-screen, which means limited resolution.
That's why we went with screen captures: if all views are in the same render window anyway, just capture the complete screen, divide it up into sections, publish those on their own topics.

We never got that far obviously.

You already participated in the discussion in https://github.com/tud-cor/FS19_modROS/issues/42. Without access to internal framebuffers (https://github.com/tud-cor/FS19_modROS/issues/22), I doubt true multi-view cameras are not going to be possible.

**Reply by gavanderhoorn** [reply]
Having written all of that: there are ways to replace embedded Lua interpreters in executables, such as the FS main executable. Doing that, in combination with editing other game files, *may* allow more straightforward use of "regular" Lua APIs, which would make at least getting data out and into FS easier.

But that would require *much* more modifications to FS than we were willing to do. As I mentioned above, `modROS` was a proof of concept, to see whether it'd make sense to spend more time on implementing a more complete integration.

Due to limitations already discussed, we decided not to invest in that. But we did want to make/keep `modROS` open source. That way others could potentially continue/contribute/benefit.

**Comment by tsbertalan** [comment]
*Ugh; I need to delete and repost my previous long response, since I added it under the wrong single-use user. Imagine that this comment came before @gavanderhoorn 's last:*

I see, the filelike pipe approach sounds like a pretty reasonable way to get high-bandwidth output then. I guess so long as you start the Python receiver first, Windows will treat it as such a special pipe file rather than a real file; I'm not familiar with such things.

Could it also do input if you had separate pipe files for both? Avoiding the memory manipulation would be desirable. I don't imagine anticheat is such a big deal for FS, but it could be nice not to need to take caution about multiplayer--it might be a natural (if expensive in hardware!) way to do multi-agent research.

    some books available ... focuses on things like UIs

Yes, they have a "for dummies" PDF, but it's all about making new vehicles and things like that. If you had any better links I'm all ears--the Giants forums seem to be [full](https://forum.giants-software.com/viewtopic.php?p=1212152&hilit=documentation+tutorial#p1212152) of [people](https://forum.giants-software.com/viewtopic.php?p=792204&hilit=documentation+tutorial) complaining about this, though. I found [this line](https://forum.giants-software.com/viewtopic.php?p=407584&hilit=documentation+tutorial#p407584) particularly funny:

    The script documentation will be released after all known bugs are fixed.

Anyway, I'll stop complaining about that.

    they'd have to be on-screen, which means limited resolution

This might be a worthwhile tradeoff--divide the screen into quadrants for different cameras and render at high base resolution. It's not clear to me whether that [multiviewport](https://gdn.giants-software.com/thread.php?categoryId=3&threadId=709) thing lets you render directly to screen space, or whether it's a render-to-texture method (for diagetic mirrors or cameras, as you say). Anyway, this is clearly a future-nice-to-have thing.

For depth/segmentation data, I commented on https://github.com/tud-cor/FS19_modROS/issues/22 about a shader post-processing approach that seemed promising to me. But that's also clearly an even-more-future-nice-to-have.

    replace embedded Lua interpreters

I didn't even consider such an approach. I guess that could be cleaner in some respects, but after your helpful walkthrough above, I'm starting to warm to the existing pipe/shared memory approach. It's not like you want to do any fancy numerical computations on the Lua side; since you want to make the modROS layer as thin as possible, there might not be too much benefit beyond sockets for an interpreter swap.

Anyway, I don't want to promise anything, but I guess this is enough information for me to start trying to port to FS22. I'll try not to bother you more, but will ask here if I have more questions beyond the above.

**Reply by gavanderhoorn** [reply]
> Could it also do input if you had separate pipe files for both? Avoiding the memory manipulation would be desirable. 

it would have been nice, if FS Lua allowed to read arbitrary data from files.

IIRC, it only supports reading XML documents. And only in DOM mode, not SAX. That limited us significantly, as you can't have an 'infinite' XML file that way.

But again, it's been a while.

> I don't imagine anticheat is such a big deal for FS, but it could be nice not to need to take caution about multiplayer--it might be a natural (if expensive in hardware!) way to do multi-agent research.

yes, we thought so too. The warning is mostly there for multi-player contexts.

> > they'd have to be on-screen, which means limited resolution
> 
> This might be a worthwhile tradeoff--divide the screen into quadrants for different cameras and render at high base resolution. It's not clear to me whether that [multiviewport](https://gdn.giants-software.com/thread.php?categoryId=3&threadId=709) thing lets you render directly to screen space, or whether it's a render-to-texture method (for diagetic mirrors or cameras, as you say). Anyway, this is clearly a future-nice-to-have thing.

exactly our thoughts as well. Hence #22.

> For depth/segmentation data, I commented on #22 about a shader post-processing approach that seemed promising to me. But that's _also_ clearly an even-more-future-nice-to-have.

The last line in the OP of #22 actually mentions reshader :)

So good idea.

> > replace embedded Lua interpreters
> 
> I didn't even consider such an approach. I guess that could be cleaner in some respects, but after your helpful walkthrough above, I'm starting to warm to the existing pipe/shared memory approach. It's not like you want to do any fancy numerical computations on the Lua side; since you want to make the modROS layer as thin as possible, there might not be too much benefit beyond sockets for an interpreter swap.

indeed. Get data in, get data out. The rest is on the ROS side. And not just because you don't have to fight the Lua sandbox any more. Doing 'interesting' things in the Lua sandbox of a game isn't exactly what I would prefer.

`modROS` is essentially a 'driver' for some virtual hw.

> Anyway, I don't want to promise anything, but I guess this is enough information for me to start trying to port to FS22. I'll try not to bother you more, but will ask here if I have more questions beyond the above.

I'd be interested in your progress.

re: port: I would expect most things to "just work". Not saying there aren't points to improve of course.

If you'd be interested, merging some PRs here might even breath new live into `modROS`

**Comment by gavanderhoorn** [comment]
@tsbertalan: just curious: have you been able to make any progress with FS22?

**Comment by matching-transforming-anns** [comment]
I decided to focus on other simulation platforms instead for now, though I
might come back to this after making progress with my core ideas on simpler
systems.

**Comment by gavanderhoorn** [comment]
@tsbertalan: I was curious about the "open pipe file directly from Lua / FS mod" part of our discussion:

@gavanderhoorn wrote:
> @tsbertalan wrote:
> > [This discussion](https://forum.giants-software.com/viewtopic.php?t=182643) on someone's "telemetry" mod seems related, though I believe [they used the same approach as you](https://github.com/Marciel032/FarmingSimulatorTelemetry/blob/7b93c844ab52695b15135f35ea558ee2e2ff8897/TelemetriaFarmingSimulator/FSTelemetry.lua#L588).
> 
> Seems so. It seems slightly less cumbersome though, as they seem to directly open the named pipe from FS. Our initial testing seemed to imply that wasn't allowed / supported, which is why we went with the symlink approach.
> 
> Would be a nice 'final' update to remove that, as it would make setting up and using `modROS` a lot easier.

<del>I'm not sure whether the author of `FarmingSimulatorTelemetry` tested his mod with the Steam version of FS19, but I can't find the pipe on my system, even if the mod is active.</del>

<del>I changed the mod's `FSTelemetry:RefreshPipe()` to check whether the pipe's fd is `nil` or not ([here](https://github.com/Marciel032/FarmingSimulatorTelemetry/blob/7b93c844ab52695b15135f35ea558ee2e2ff8897/TelemetriaFarmingSimulator/FSTelemetry.lua#L583-L586)), and it appears it always is. I'm thinking this is the same issue we ran into when we initially tested some implementation strategies for `modROS` and which made us go with the symlink-to-a-named-pipe approach.</del>

<del>Just to make sure I haven't overlooked anything I'm going to try some other approaches, but I'm not hopeful at this point directly opening the pipe is going to work.</del>

See comment below.

**Reply by gavanderhoorn** [reply]
Ignore all of that. Turns out I was / we were mistaken. It is actually possible to directly open named pipes from FS Lua, similar to how the FSTelemetry mod does it.

Not sure what we missed in our initial tests, but see #58 for the `modROS` side to this.

Thanks for drawing my attention to this.

**Comment by matching-transforming-anns** [comment]
Hey happy to help where I can. I just started a new job and don’t have time to play with this project right now, but I’ll make a note for, say, July, to see if things have changed for me.

 

Then I guess you can close this ill-defined “issue”, unless you want it to be a request for a roadmap (which, going by your previous comment about “final update”, might not be of interest).
